### PR TITLE
dd: handle stdout redirected to seekable file

### LINKED ISSUE #3542: dd: writes null bytes to device if stdout is redirect to device
@jfinkels I took the liberty to open a new issue with a summary for all the findings you published in https://github.com/uutils/coreutils/issues/3447

`gnu/tests/dd/skip-seek-past-dev.sh` runs
```bash
device=$(df --output=source . | tail -n1)
dev_size=$(blockdev --getsize64 $device)
DEV_OFLOW=$(expr $dev_size + 1)
dd bs=1 seek=$DEV_OFLOW count=0 status=noxfer > "$device"
```
and expects `$?`=1 and stderr=`dd: 'standard output': cannot seek: Invalid argument`.

However, uu_tail does not check if `stdout` is a redirect to a seekable device and then writes null bytes to `$device` !
https://github.com/uutils/coreutils/blob/d894847fc69bcc2e9901b72de689a84cf24837dc/src/uu/dd/src/dd.rs#L301-L309

### possible fix
Haven't looked into it much but a possible fix could be to get the redirect for `/dev/stdout` similar to this:
https://github.com/uutils/coreutils/blob/d894847fc69bcc2e9901b72de689a84cf24837dc/src/uu/stat/src/stat.rs#L558-L562
and then apply the seek checks for the new path.

### PR DESCRIPTION
**Pull request by jfinkels**
This pull request fixes a bug in `dd` where null bytes would be unintentionally written if stdout were redirected to a seekable file. For example, before this commit, if `dd` were invoked from the command-line as

    dd if=infile bs=1 count=10 seek=5 > /dev/sda1

then five zeros would be written to `/dev/sda1` before copying ten bytes of `infile` to `/dev/sda1`. After this commit, `dd` will correctly seek five bytes forward in `/dev/sda1` before copying the ten bytes of `infile`. I wasn't sure how to write a test for this.

In order to accomplish this, the first commit moves the argument parsing out of the `Input::new()` and `Output::new()` functions. This was needed in order to be able to determine whether we need to use the `Output<File>` specialization even when no outfile is specified in the command-line arguments. The code is a bit messy, but I think it should work.

Fixes #3542.

### CODE REVIEW

File: src/uu/dd/src/dd.rs [hunk 1]
```diff
@@ -703,6 +683,49 @@ fn append_dashes_if_not_present(mut acc: Vec<String>, mut s: String) -> Vec<Stri
...
+    match File::open(p) {
+        Ok(mut f) => {
+            f.seek(SeekFrom::Current(0)).is_ok()
+                && f.seek(SeekFrom::End(0)).is_ok()
+                && f.seek(SeekFrom::Start(0)).is_ok()
+        }
+        Err(_) => false,
+    }
+}
+
```
Comment (jhscheer) [inline src/uu/dd/src/dd.rs:728]: There's similar functionality in other uutils, e.g. tail. We should think about a follow-up PR to move this kind of checks to a central file to reduce code duplication.

**Comment by jhscheer** [comment]
I know you had serious problems with this on your system before, so sorry for asking. Did you try to run `tests/dd/skip-seek-past-dev.sh` as root with the changes from this PR?

**Comment by jfinkels** [comment]
I did not try to run that test yet. I thought the CI system executes that test because of the issues we had in #3442, but now it looks like the test is skipped? From the logs:
```
2022-08-27T21:23:35.8791953Z skip-seek-past-dev.sh: skipped test: must be run as root
2022-08-27T21:23:35.8830694Z SKIP: tests/dd/skip-seek-past-dev.sh
```

**Comment by jhscheer** [comment]
Yes, sadly the CI doesn't run any GNU tests which requires root. That's why I asked.

**Comment by jfinkels** [comment]
Okay, I'll install a virtual machine and try it there.

**Comment by jfinkels** [comment]
I checked out the branch from pull request #3882 (which includes the changes from this branch as well as the change to `df` that triggered the issue before) and tested by running
```
sudo bash util/run-gnu-test.sh tests/dd/skip-seek-past-dev.sh
```
the relevant part of the error log is:
```
--- err_ok	2022-08-30 19:04:36.700725532 -0400
+++ err	2022-08-30 19:04:36.596723639 -0400
@@ -1,3 +1 @@
-dd: 'standard output': cannot seek: Invalid argument
-0+0 records in
-0+0 records out
+dd: failed to seek in output file: Invalid input
FAIL tests/dd/skip-seek-past-dev.sh (exit status: 1)
```
The message "failed to seek in output file: Invalid input" comes from https://github.com/uutils/coreutils/blob/282774a3ef1c6ef38aa22d0fa2271f3971d3a00d/src/uu/dd/src/dd.rs#L543-L544 which is in the implementation for `Output<File>`. That's what I expected, so that's good!

**Pull request merged**